revisions/git/tags.go

```diff
@@ -0,0 +1,157 @@
+
+func (rcs refCommits) Len() int {
+	return len(rcs)
+}
+func (rcs refCommits) Swap(i, j int) {
+	rcs[i], rcs[j] = rcs[j], rcs[i]
+}
+func (rcs refCommits) Less(i, j int) bool {
+	if rcs[i].commit == nil {
+		return false
+	}
+	if rcs[j].commit == nil {
+		return true
+	}
+	return rcs[i].commit.Committer.When.After(rcs[j].commit.Committer.When)
+}
+
+func NewTimeOrderedReferenceIter(iter storer.ReferenceIter, repo Repository) (storer.ReferenceIter, error) {
+	rcs := make([]refCommit, 0)
+	var ref *plumbing.Reference
+	var err error
+	for ref, err = iter.Next(); ref != nil && err == nil; ref, err = iter.Next() {
+		commit, err := repo.CommitObject(ref.Hash())
+		if err != nil {
+			log.Printf("WARN: Failed to lookup commit for reference %v", ref)
+			continue
+		}
+		rcs = append(rcs, refCommit{
+			ref,
+			commit,
+		})
+	}
+	if err != io.EOF {
+		return nil, err
+	}
+	iter.Close()
+	sort.Sort(refCommits(rcs))
+	refs := make([]*plumbing.Reference, 0, len(rcs))
+	for _, rcs := range rcs {
+		refs = append(refs, rcs.ref)
+	}
+	return storer.NewReferenceSliceIter(refs), nil
+}
+
+func NewMergedPRIter(iter storer.ReferenceIter, repo Repository) (storer.ReferenceIter, error) {
+	iter, err := NewTimeOrderedReferenceIter(storer.NewReferenceFilteredIter(func(ref *plumbing.Reference) bool {
+		if ref == nil {
+			return false
+		}
+		return strings.HasPrefix(string(ref.Name()), "refs/tags/merge_pr_")
+	}, iter), repo)
+	if err != nil {
+		log.Printf("ERRO: Failed to construct new merged PR iter: %v", err)
+		return nil, err
+	}
+	return iter, err
+}
+
+type StartReferenceIter struct {
+	startAt func(ref *plumbing.Reference) bool
+	iter    storer.ReferenceIter
+	started bool
+}
+
+func (iter *StartReferenceIter) Next() (ref *plumbing.Reference, err error) {
+	if iter.started {
+		return iter.iter.Next()
+	}
+	for ref, err = iter.iter.Next(); err == nil; ref, err = iter.iter.Next() {
+		if iter.startAt(ref) {
+			iter.started = true
+			return ref, nil
+		}
+	}
+	return ref, err
+}
+
+func (iter *StartReferenceIter) ForEach(f func(*plumbing.Reference) error) error {
+	return iter.iter.ForEach(func(ref *plumbing.Reference) error {
+		if iter.started {
+			return f(ref)
+		}
+		if iter.startAt(ref) {
+			iter.started = true
+			return f(ref)
+		}
+		return nil
+	})
+}
+
+func (iter *StartReferenceIter) Close() {
+	iter.iter.Close()
+}
+
+func NewStartReferenceIter(iter storer.ReferenceIter, startAt func(ref *plumbing.Reference) bool) storer.ReferenceIter {
+	return &StartReferenceIter{
+		startAt,
+		iter,
+		false,
+	}
+}
+
+type StopReferenceIter struct {
+	stopAt func(ref *plumbing.Reference) bool
+	iter   storer.ReferenceIter
+}
+
+func (iter StopReferenceIter) Next() (ref *plumbing.Reference, err error) {
+	ref, err = iter.iter.Next()
+	if err != nil {
+		return ref, err
+	}
+	if iter.stopAt(ref) {
+		iter.Close()
+		return nil, io.EOF
+	}
+	return ref, err
+}
+
+func (iter StopReferenceIter) ForEach(f func(*plumbing.Reference) error) error {
```

Comment:
Is there something inefficient about iterating all of the refs? Or are you just following the pattern `go-git` uses?
FWIW I've seen abuse of a `chan` so that `range` can be used fairly often. Would look something like

    func (iter StopReferenceIter) Iterate() chan *plumbing.Reference {
      c := make(chan *plumbing.Reference)
      go func() {
        iter.iter.ForEach(f func(v *plumbing.Reference) error {
          c <- v
        })
        close(c)
      }()
      return c
    }

Reply:
I created `StopReferenceIter` so that I have a consistent interface for expressing time-bounded iteration at the public API layer. I generally agree that `Next()`/`ForEach()` is not generally the idiomatic way in golang, but (as you mentioned) it's what `go-git` uses, so I just ran with that. I implement (but don't actually use in my API code) `ForEach()` specifically so that `StopReferenceIter` implements `plumbing.ReferenceIter`.